src/model/entities/segment.ts

```diff
@@ -88,7 +109,51 @@ export class Segment {
   }
 
   public setParts(parts: Part[]): void {
-    this.parts = parts.sort((partOne: Part, partTwo: Part) => partOne.rank - partTwo.rank)
+    this.parts = parts.sort(this.compareParts)
+  }
+
+  private compareParts(partOne: Part, partTwo: Part): number {
+    return partOne.getRank() - partTwo.getRank()
+  }
+
+  public addPart(part: Part): void {
+    if (!this.isInsertingPartAllowed(part)) {
+      return
+    }
+
+    const doesPartAlreadyExistOnSegment: boolean = this.parts.some(p => p.id === part.id)
```

Comment:
Suggestion: You could rename the parameter to `partToAdd`. That way, you can avoid using `p` as an abbreviation for part.

Reply:
Updated